presto-tests/src/test/java/com/facebook/presto/memory/TestClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,117 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.TestQueryRunnerUtil;
+import com.facebook.presto.server.testing.TestingPrestoServer;
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.tests.DistributedQueryRunner;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.SettableFuture;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.ExecutionException;
+
+import static com.facebook.presto.execution.QueryState.FAILED;
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.facebook.presto.execution.TestQueryRunnerUtil.cancelQuery;
+import static com.facebook.presto.execution.TestQueryRunnerUtil.createQuery;
+import static com.facebook.presto.execution.TestQueryRunnerUtil.waitForQueryState;
+import static com.facebook.presto.memory.LocalMemoryManager.GENERAL_POOL;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static org.testng.Assert.assertEquals;
+
+@Test(singleThreaded = true)
+public class TestClusterMemoryLeakDetector
+{
+    @Test(timeOut = 90_000)
+    public void testLeakDetector()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = TestQueryRunnerUtil.createQueryRunner()) {
+            TestingPrestoServer worker = queryRunner.getServers().get(0);
+            MemoryPool generalPool = worker.getLocalMemoryManager().getPool(LocalMemoryManager.GENERAL_POOL);
+            ClusterMemoryLeakDetector leakDetector = queryRunner.getCoordinator().getClusterMemoryLeakDetector();
+
+            Session testSession = testSessionBuilder()
+                    .setCatalog("tpch")
+                    .setSchema("sf100000")
+                    .setSource("test")
+                    .setClientTags(ImmutableSet.of())
+                    .setResourceEstimates(null)
+                    .build();
+
+            // start a long running query
+            QueryId queryId = createQuery(queryRunner, testSession, "SELECT COUNT(*) FROM lineitem");
+
+            // reserve some memory, the query won't be able to release it so effectively this is a leak
+            generalPool.reserve(queryId, 1L);
+
+            // wait for the query to start
+            waitForQueryState(queryRunner, queryId, RUNNING);
+
+            // wait until the cluster-level memory pools are updated
+            waitUntilQueryMemoryReservationIsObserved(queryRunner, queryId);
```

Comment:
could cluster memory manager be updated before listener is added in `waitUntilQueryMemoryReservationIsObserved`?

Reply:
It can, but it will get updated again as the updates are periodic: `SqlQueryManager` has a scheduled thread that calls `ClusterMemoryManager::process()`, which calls `ClusterMemoryManager::updatePools()`, and it calls these listeners.